### PR TITLE
chore: avoiding calling through the mesh to the proxy

### PR DESCRIPTION
**Pull request by franciscolopezsancho**
This is the first attempt to avoid calling the proxy through DNS. 

Maybe the better way to do this is to change https://github.com/lightbend/kalix-proxy/blob/9215b87b92eb46f6fae443c3ef99468130af360c/proxy/core/src/main/resources/reference.conf#L5-L6

and add only one param which is the port kind of 
```
  proxy-port = 9000
  proxy-port = ${?KALIX_PROXY_PORT}
```
and then consume that from this SDK here https://github.com/lightbend/kalix-jvm-sdk/blob/1a00df65e569679804502124a67fdab2988b876d/sdk/java-sdk/src/main/scala/kalix/javasdk/impl/GrpcClients.scala#L109

For that I would need to deal with the deployment yaml files which so far I don't know how to test. 
Do you think this could be the first part of a two-step solution?

References https://github.com/lightbend/kalix-proxy/issues/1469

### CODE REVIEW

File: sdk/java-sdk/src/main/scala/kalix/javasdk/impl/GrpcClients.scala
Comment (johanandren) [inline sdk/java-sdk/src/main/scala/kalix/javasdk/impl/GrpcClients.scala:114]: I think it would be better if this decision was moved to the proxy and handed to the SDK through the `ProxyInfo` message (defined in discovery.proto in the proxy)

Comment (johanandren) [reply sdk/java-sdk/src/main/scala/kalix/javasdk/impl/GrpcClients.scala:114]: (just like hos the proxyHostName is today)

Comment (franciscolopezsancho) [reply sdk/java-sdk/src/main/scala/kalix/javasdk/impl/GrpcClients.scala:114]: Didn't see that coming. Thanks!

**Comment by franciscolopezsancho** [comment]
Depends on https://github.com/lightbend/kalix-proxy/pull/1599

File: sdk/java-sdk/src/main/scala/kalix/javasdk/impl/GrpcClients.scala
Comment (franciscolopezsancho) [inline sdk/java-sdk/src/main/scala/kalix/javasdk/impl/GrpcClients.scala:111]: `host` should be always `localhost`

File: sdk/java-sdk/src/main/scala/kalix/javasdk/impl/DiscoveryImpl.scala
Comment (octonato) [inline sdk/java-sdk/src/main/scala/kalix/javasdk/impl/DiscoveryImpl.scala:74]: "this version is probe" is not something that users will understand. I don't understand it myself. :-)

I think we emit a log similar to what we are logging in the `else` branch of this `if`.

**Review by octonato — APPROVED** [review]
LGTM, with a minor comment about logging

**Comment by franciscolopezsancho** [comment]
Before we merge this PR, [this](https://github.com/lightbend/kalix-proxy/pull/1599) needs to be released

File: sdk/java-sdk/src/main/scala/kalix/javasdk/impl/DiscoveryImpl.scala
Comment (octonato) [inline sdk/java-sdk/src/main/scala/kalix/javasdk/impl/DiscoveryImpl.scala:80]: 👍

File: sdk/java-sdk/src/main/scala/kalix/javasdk/impl/GrpcClients.scala
Comment (octonato) [inline sdk/java-sdk/src/main/scala/kalix/javasdk/impl/GrpcClients.scala:113]: As discussed on the stand-up today, we still need to support the old code because a user can update their SDK version, but still have the docker-compose using an old version of the proxy. 

When that happens, users will probably revert their SDK version believing that the SDK is broken. It will only broken in dev mode though. 

So, we need to add two cases to this pattern matching.

```suggestion
    (proxyHostname, proxyPort) match {
      case (Some(hostname), Some(port)) =>
        getGrpcClient(serviceClass, hostname, port, localAddHeader)
      case (Some("localhost"), None) => getGrpcClient(serviceClass, "localhost", selfPort.getOrElse(9000), localAddHeader)
      case (Some(selfName), None)    => getGrpcClient(serviceClass, selfName, 80, localAddHeader)
      case _ =>
        throw new IllegalStateException(
          "Service proxy hostname and port are not set by proxy at discovery, too old proxy version?")
```

Basically, if port is not set (old proxy behavior), we fallback to the old strategy.

Same for the javascript sdk

File: sdk/java-sdk/src/main/scala/kalix/javasdk/impl/DiscoveryImpl.scala
Comment (octonato) [inline sdk/java-sdk/src/main/scala/kalix/javasdk/impl/DiscoveryImpl.scala:97]:
```suggestion
      if (in.internalProxyHostname.isEmpty) {
```

**Review by octonato — APPROVED** [review]
LGTM

We still need to wait for a proxy release to get this merged. We are depending on the new protocol file that brings in `internalProxyHostname`

**Comment by octonato** [comment]
Commits are squashed and then rebased on top of #1061. 

We should merge this one after #1061 and before releasing.

**Comment by octonato** [comment]
This one is ready for merge. Will do when green

**Pull request merged**
